controlpanel/api/models/tool.py

```diff
@@ -0,0 +1,131 @@
+    name = models.CharField(max_length=100, blank=False)
+    values = JSONField(default=dict)
+    version = models.CharField(max_length=100, blank=False)
+
+    class Meta(TimeStampedModel.Meta):
+        db_table = "control_panel_api_tool"
+        ordering = ('name',)
+
+    def __repr__(self):
+        return f'<Tool: {self.chart_name} {self.version}>'
+
+    def url(self, user):
+        return f'https://{user.slug}-{self.chart_name}.{settings.TOOLS_DOMAIN}/'
+
+
+class ToolDeploymentManager:
+    """
+    Emulates a Django model manager
+    """
+
+    def create(self, *args, **kwargs):
+        tool_deployment = ToolDeployment(*args, **kwargs)
+        tool_deployment.save()
+        return tool_deployment
+
+    def filter(self, **kwargs):
+        user = kwargs.get('user')
+        filter = Q(chart_name=None)  # Always False
+        for deployment in cluster.list_tool_deployments(user):
+            _, version = deployment.metadata.labels["chart"].rsplit("-", 1)
+            filter = filter | (
+                Q(chart_name=deployment.metadata.labels["app"]) &
+                Q(version=version)
+            )
+
+        tools = Tool.objects.filter(filter)
+        return [ToolDeployment(tool, user) for tool in tools]
+
+
+class ToolDeployment:
+    """
+    Represents a deployed Tool in the cluster
+    """
+
+    DoesNotExist = django.core.exceptions.ObjectDoesNotExist
+    Error = cluster.ToolDeploymentError
+    MultipleObjectsReturned = django.core.exceptions.MultipleObjectsReturned
+
+    objects = ToolDeploymentManager()
+
+    def __init__(self, tool, user):
+        self._subprocess = None
+        self.tool = tool
+        self.user = user
+
+    def __repr__(self):
+        return f'<ToolDeployment: {self.tool!r} {self.user!r}>'
+
+    def delete(self):
+        """
+        Remove the release from the cluster
+        """
+        cluster.delete_tool_deployment(self)
+
+    @property
+    def host(self):
+        return f'{self.user.slug}-{self.tool.chart_name}.{settings.TOOLS_DOMAIN}'
+
+    def save(self, *args, **kwargs):
+        """
+        Deploy the tool to the cluster (asynchronous)
+        """
+        self._subprocess = cluster.deploy_tool(self.tool, self.user)
+
+    @property
+    def status(self):
+        """
+        Get the current status of the deployment.
+        Polls the subprocess if running, otherwise returns idled status.
+        """
+        if self._subprocess:
+            # poll subprocess and maybe parse any output to get status
+            status = self._poll()
+            if status:
+                return status
+
+        return cluster.get_tool_deployment_status(self)
+
+    def _poll(self):
+        """
+        Poll the deployment subprocess for status
+        """
+
+        if self._subprocess.poll() is None:
+            return cluster.TOOL_DEPLOYING
+
+        if self._subprocess.returncode:
+            log.error(err_msg)
```

Comment:
FYI: I'm getting `name 'err_msg' is not defined` from this as `err_msg` is not defined (apologies if you were already aware of this).

Reply:
Played with this - I'm not familiar with running subprocess in Python at all, replace with `log.error(self._subprocess.returncode)` prints `ERROR Failed deploying RStudio for xoen` which seems very weird to me as I thought a process's return code was a number not a string 😁